### DISCUSSION TITLE
Add a new configuration option to toggle new/old behavior to warn about incomplete implementation of abstract classes

### DISCUSSION THREAD

**Discussion by minmax** [microsoft/pyright#7622] [General]
Previously (some releases ago) pyright warn me in class definition line if class does't implement all abstract method.
But then this behavior was changed in favor of compatibility with mypy, and pyright now show errors only for class creation lines.

current behavior:

```python
class Abstract(Protocol):
    def foo(self) -> None:
        ...


class Sub(Abstract):
    pass


Sub() # reportAbstractUsage
```

and previous

```python
class Abstract(Protocol):
    def foo(self) -> None:
        ...


class Sub(Abstract): # reportAbstract/CommonError
    pass
```

I'm developing a library that itself creates class instances, i.e. there will be no explicit call to MyClass() anywhere in the code. Which completely makes it impossible for me to detect errors of this type. But if it were possible to return the “old” error reporting mode, it would help tremendously.

e.g.

```python
T = TypeVar("T")


class A:
    pass


class B:
    a: A


def magic(call: Callable[..., T]) -> T: ...


b: B = magic(B)

assert isinstance(b.a, A)
```

**Comment by erictraut** [comment]
The new behavior is compliant with the typing spec whereas the old behavior was not, so it's unlikely that we'd add back the old behavior.

One potential workaround is to mark the child class as `@final` (i.e. use the `typing.final` decorator on the class). Pyright will then report any unimplemented methods for the child class. This workaround applies only if you don't intend to further subclass those classes.

Another option is to use explicit instantiation in your test cases. You can then run pyright on your test cases to statically catch any unimplemented methods.

**Reply by minmax** [reply]
Thanks for the answer, `@final` is really what I need. And for those cases when it does not work, tests will do the work.
